Split path identifiers as odd-even for client and server initiated (#329)
quicwg/multipath: draft-ietf-quic-multipath.md

means the current Connection ID can only be used on the corresponding path.

Sequence Number:
The sequence number assigned to the connection ID by the sender on the path
specified by Path Identifier, encoded as a variable-length integer.

## MAX_CLIENT_PATHS and MAX_SERVER_PATHS frames {#max-paths-frame}

The MAX_CLIENT_PATHS (type=0x15228c0c) and MAX_SERVER_PATHS (type=0x15228c0d) frames inform the peer of the cumulative
number of client initiated and server initiated paths for which it MAY publish
MP_NEW_CONNECTION_ID frames.

MAX_PATHS frames are formatted as shown in {{fig-max-paths-frame-format}}.

~~~
MAX_CLIENT_PATHS Frame {
  Type (i) = 0x15228c0c,
  Maximum Number of Client Paths (i),
}

MAX_SERVER_PATHS Frame {
  Type (i) = 0x15228c0d,
   Maximum Number of Server Paths (i),
}

~~~
{: #fig-max-paths-frame-format title="MAX_PATHS Frame Format"}

MAX_CLIENT_PATHS and MAX_SERVER_PATHS frames contain the following field:

Maximum Number of Client/Server Path:

> : A count of the cumulative number of respectively client orserver path that can be opened

Reviewer: ```suggestion
: A count of the cumulative number of respectively client-initiated or server-initiated paths that can be opened
```

over the lifetime of the connection. The client path identifiers sent by either
endpoint MUST be lower than 2 times the Maximum Number of Client Paths announced py the peer.
The server path identifiers sent by either
endpoint MUST be lower than 2 times the Maximum Number of Server Paths announced py the peer

> plus 1. These values cannot exceed 2^30-1, as it is not

Reviewer: This should be 2^31 - 1, right? (2^31 - 1) * 2 = 2^32 - 2 (max client-initiated path ID) and (2^31 - 1) * 2 + 1 =  2^32 - 1 (max client-server path ID)
```suggestion
plus 1. These values cannot exceed 2^31-1, as it is not
```

possible to encode Path IDs larger than 2^32-1.

Receipt of a frame that permits
opening of a path with Path Identifier larger than the corresponding limit MUST be treated

> as a connection error of type FRAME_ENCODING_ERROR. This MUST be
> an even value for client initiated paths, and an odd value for
> server initiated path.

Reviewer: I think the aforementioned commit resolves the issue, modulo the removal of the old text referring to old frame versions.
```suggestion
as a connection error of type FRAME_ENCODING_ERROR.
```

specified by Path Identifier, encoded as a variable-length integer. 

## MAX_CLIENT_PATHS and MAX_SERVER_PATHS frames {#max-paths-frame}

The MAX_CLIENT_PATHS and MAX_SERVER_PATHS frame (type=0x15228c0b) informs the peer of the cumulative
number of client initiated and server initiated paths for which it MAY publish
MP_NEW_CONNECTION_ID frames. 

MAX_PATHS frames are formatted as shown in {{fig-max-paths-frame-format}}.

~~~
MAX_CLIENT_PATHS Frame {
  Type (i) = 0x15228c0c,
  Maximum Path Identifier (i),
}

MAX_SERVER_PATHS Frame {
  Type (i) = 0x15228c0d,
  Maximum Path Identifier (i),
}

~~~
{: #fig-max-paths-frame-format title="MAX_PATHS Frame Format"}

MAX_CLIENT_PATHS and MAX_SERVER_PATHS frames contain the following field:

Maximum Path Identifier:
: A count of the cumulative number of path that can be opened
over the lifetime of the connection. This value cannot exceed 2^32-1, as it is not
possible to encode Path IDs larger than 2^32-1. Receipt of a frame that permits
opening of a path with Path Identifier larger than this limit MUST be treated
as a connection error of type FRAME_ENCODING_ERROR. This MUST be
an even value for client initiated paths, and an odd value for

> server initiated path.

Author: Yes, we could do that. The logical solution would be to carry a "number" instead of an identifier, and to have a derivation function, such as 2*N for client paths and 2*N+1 for server paths. But I think we need to think a bit more. The purpose of the "MAX_PATHS" is to control the amount of resource dedicated to path managements. The resource comes in two modes:

* The number of MP_CONNECTION_ID that the endpoint has to remember,
* The number of path contexts that the endpoint will handle.

I think that the main purpose of the MAX_PATHS is to limit the number of MP_CONNECTION_ID. The peer can only create a new path if it has received and sent an MP_CONNECTION_ID for that path. Endpoints can control the creation of paths by limiting the creation of MP_CONNECTION_ID, which is a local decision. But endpoints need to also control the number of MP_CONNECTION_ID that the peer can send -- otherwise, the peer could announce that it supports 2^31 paths, and send as many connection ID. That would blow the memory allocation.

Controlling a "max number of paths" is not sufficient. The malevolent peer could create lots of MP_CONNECTION_ID, then issue PATH_ABANDON, the create more of that. I am pretty sure that repeating that cycle many times could lead to a DOS attack. Also, we understand that "number of paths" is imprecise, because of transmission delays, packet loss, etc. It is simpler to have a "MAX_PATHS" value that state "no path identifier larger than N" (or 2*N+1). Once the number of connection ID is limited, the peers can control how much resource will be used.

If we agree on that, we also need to decide what "MAX_SERVER_PATH=0" means. Assume we agree that N=0 corresponds to path_ID=2*N+1 = 1. Does stating "MAX_SERVER_PATH=0" means that "path 1 is OK", or does it mean "path 1 is not OK, because 1 >= 2*0 +1" ?

> server initiated path.

Author: @qdeconinck I think that the commit [reworking the max path frames](https://github.com/quicwg/multipath/pull/329/commits/46fb9fa7728789ab1271836c4d8562e8b4e755a8) fixes your issue. Please mark as resolved if it does.

> server initiated path.

Reviewer: Thanks for the reworked text! Modulo my suggestion to remove the trailing text, I think it is better now.

Only one point that we may need to discuss now is the meaning of MAX_SERVER_PATHS=0, but also MAX_CLIENT_PATHS=0. I think MAX_CLIENT_PATHS=0 should actually allow Path ID 0, while MAX_SERVER_PATHS=0 should also enable Path ID 1. I see two reasons for this. First, it would allow using the whole space of the identifiers, without needing to cope with an invalid 0 value -- Path ID 0 is always valid as it is the initial path of the connection. Second, MAX_CLIENT_PATHS frames only appear when multipath is negotiated.

Side note related to this, I think the MAX_SERVER_PATHS should only appear when the initial_max_server_paths has been advertised by both endpoints. Otherwise, this should be a PROTOCOL_VIOLATION.

> server initiated path.

Author: Yes, I struggled with that. if we add the text that "MAX_SERVER_PATHS should only appear when the initial_max_server_paths has been advertised by both endpoints", then "0 allows 1 on server" makes sense.

Of course, if we decide to remove the server capability, then this becomes easier...

Loss or reordering can cause an endpoint to receive a MAX_CLIENT_PATHS or MAX_SERVER_PATHS frame with
a lower path limit than was previously received. MAX_CLIENT_PATHS and MAX_SERVER_PATHS frames that
do not increase the path limit MUST be ignored.

An endpoint MUST NOT initiate publish MP_NEW_CONNECTION_ID frames with a path ID higher
than the Maximum Paths value advertised in MAX_CLIENT_PATHS for even numbered paths

> or in MAX_SERVER_PATHS for odd numbers.

Author: See discussion above. we need to agree on principles before fixing the text.

> or in MAX_SERVER_PATHS for odd numbers.

Author: I don't like the "cumulative number of paths" because it is ambiguous -- there are many cases in which the apparent number of paths is not the same at both ends, e.g., if probes are in progress or if abandon is in transit. "MAX_STREAMS" get around that by forcing an automatic opening of streams -- for example, if the only stream is stream 0 and the server receives something for stream 16, then streams 4, 8 and 12 are automatically opened.
